// supi/lib/storeStatus/extendStoreStatusOptions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("@pnpm/utils");
const path = require("path");
const defaults = async (opts) => {
    const prefix = opts.prefix || process.cwd();
    const lockfileDirectory = opts.lockfileDirectory || prefix;
    return {
        bin: path.join(prefix, 'node_modules', '.bin'),
        force: false,
        forceSharedLockfile: false,
        independentLeaves: false,
        lockfileDirectory,
        prefix,
        registries: utils_1.DEFAULT_REGISTRIES,
        shamefullyFlatten: false,
        store: opts.store,
        useLockfile: true,
    };
};
exports.default = async (opts) => {
    if (opts) {
        for (const key in opts) {
            if (opts[key] === undefined) {
                delete opts[key];
            }
        }
    }
    const defaultOpts = await defaults(opts);
    const extendedOpts = Object.assign({}, defaultOpts, opts, { store: defaultOpts.store });
    extendedOpts.registries = utils_1.normalizeRegistries(extendedOpts.registries);
    return extendedOpts;
};
