// supi/lib/install/extendInstallOptions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const constants_1 = require("@pnpm/constants");
const utils_1 = require("@pnpm/utils");
const path = require("path");
const pnpmPkgJson_1 = require("../pnpmPkgJson");
const defaults = async (opts) => {
    const packageManager = opts.packageManager || {
        name: pnpmPkgJson_1.default.name,
        version: pnpmPkgJson_1.default.version,
    };
    return {
        childConcurrency: 5,
        depth: 0,
        engineStrict: false,
        force: false,
        forceSharedLockfile: false,
        frozenLockfile: false,
        hooks: {},
        ignoreCurrentPrefs: false,
        ignoreScripts: false,
        include: {
            dependencies: true,
            devDependencies: true,
            optionalDependencies: true,
        },
        independentLeaves: false,
        localPackages: {},
        lock: true,
        lockfileDirectory: opts.lockfileDirectory || opts.prefix || process.cwd(),
        lockfileOnly: false,
        locks: path.join(opts.store, '_locks'),
        lockStaleDuration: 5 * 60 * 1000,
        nodeVersion: process.version,
        ownLifecycleHooksStdio: 'inherit',
        packageManager,
        preferFrozenLockfile: true,
        pruneLockfileImporters: false,
        pruneStore: false,
        rawNpmConfig: {},
        registries: utils_1.DEFAULT_REGISTRIES,
        repeatInstallDepth: -1,
        resolutionStrategy: 'fast',
        shamefullyFlatten: false,
        sideEffectsCacheRead: false,
        sideEffectsCacheWrite: false,
        store: opts.store,
        storeController: opts.storeController,
        strictPeerDependencies: false,
        tag: 'latest',
        unsafePerm: process.platform === 'win32' ||
            process.platform === 'cygwin' ||
            !(process.getuid && process.setuid &&
                process.getgid && process.setgid) ||
            process.getuid() !== 0,
        update: false,
        useLockfile: true,
        userAgent: `${packageManager.name}/${packageManager.version} npm/? node/${process.version} ${process.platform} ${process.arch}`,
        verifyStoreIntegrity: true,
    };
};
exports.default = async (opts) => {
    if (opts) {
        for (const key in opts) {
            if (opts[key] === undefined) {
                delete opts[key];
            }
        }
    }
    const defaultOpts = await defaults(opts);
    const extendedOpts = Object.assign({}, defaultOpts, opts, { store: defaultOpts.store });
    if (!extendedOpts.useLockfile && extendedOpts.lockfileOnly) {
        const err = new Error(`Cannot generate a ${constants_1.WANTED_LOCKFILE} because lockfile is set to false`);
        err['code'] = 'ERR_PNPM_CONFIG_CONFLICT_LOCKFILE_ONLY_WITH_NO_LOCKFILE'; // tslint:disable-line:no-string-literal
        throw err;
    }
    if (extendedOpts.userAgent.startsWith('npm/')) {
        extendedOpts.userAgent = `${extendedOpts.packageManager.name}/${extendedOpts.packageManager.version} ${extendedOpts.userAgent}`;
    }
    extendedOpts.registries = utils_1.normalizeRegistries(extendedOpts.registries);
    extendedOpts.rawNpmConfig['registry'] = extendedOpts.registries.default; // tslint:disable-line:no-string-literal
    return extendedOpts;
};
