// @pnpm/package-is-installable/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_loggers_1 = require("@pnpm/core-loggers");
const checkEngine_1 = require("./checkEngine");
const checkPlatform_1 = require("./checkPlatform");
function packageIsInstallable(pkgId, pkg, options) {
    const warn = checkPlatform_1.default(pkgId, {
        cpu: pkg.cpu || ['any'],
        os: pkg.os || ['any'],
    })
        || pkg.engines && checkEngine_1.default(pkgId, pkg.engines, {
            node: options.nodeVersion,
            pnpm: options.pnpmVersion,
        });
    if (!warn)
        return true;
    core_loggers_1.installCheckLogger.warn({
        message: warn.message,
        prefix: options.prefix,
    });
    if (options.optional) {
        core_loggers_1.skippedOptionalDependencyLogger.debug({
            details: warn.toString(),
            package: {
                id: pkgId,
                name: pkg.name,
                version: pkg.version,
            },
            prefix: options.prefix,
            reason: warn.code === 'ERR_PNPM_UNSUPPORTED_ENGINE' ? 'unsupported_engine' : 'unsupported_platform',
        });
        return false;
    }
    if (options.engineStrict)
        throw warn;
    return null;
}
exports.default = packageIsInstallable;
