// @pnpm/tarball-fetcher/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const logger_1 = require("@pnpm/logger");
const getCredentialsByURI = require("credentials-by-uri");
const mem_1 = require("mem");
const fs = require("mz/fs");
const path = require("path");
const pathTemp = require("path-temp");
const ssri = require("ssri");
const unpackStream = require("unpack-stream");
const createDownloader_1 = require("./createDownloader");
const errorTypes_1 = require("./errorTypes");
function default_1(opts) {
    const download = createDownloader_1.default({
        alwaysAuth: opts.alwaysAuth || false,
        ca: opts.ca,
        cert: opts.cert,
        fsIsCaseSensitive: typeof opts.fsIsCaseSensitive === 'boolean'
            ? opts.fsIsCaseSensitive
            : true,
        key: opts.key,
        localAddress: opts.localAddress,
        proxy: opts.httpsProxy || opts.proxy,
        registry: opts.registry,
        retry: {
            factor: opts.fetchRetryFactor,
            maxTimeout: opts.fetchRetryMaxtimeout,
            minTimeout: opts.fetchRetryMintimeout,
            retries: opts.fetchRetries,
        },
        // TODO: cover with tests this option
        // https://github.com/pnpm/pnpm/issues/1062
        strictSSL: typeof opts.strictSsl === 'boolean'
            ? opts.strictSsl
            : true,
        userAgent: opts.userAgent,
    });
    return {
        tarball: fetchFromTarball.bind(null, {
            fetchFromRemoteTarball: fetchFromRemoteTarball.bind(null, {
                download,
                getCredentialsByURI: mem_1.default((registry) => getCredentialsByURI(registry, opts.rawNpmConfig)),
                ignoreFile: opts.ignoreFile,
                offline: opts.offline,
            }),
            ignore: opts.ignoreFile,
        }),
    };
}
exports.default = default_1;
function fetchFromTarball(ctx, resolution, target, opts) {
    if (resolution.tarball.startsWith('file:')) {
        const tarball = path.join(opts.prefix, resolution.tarball.slice(5));
        return fetchFromLocalTarball(tarball, target, {
            ignore: ctx.ignore,
            integrity: resolution.integrity,
        });
    }
    return ctx.fetchFromRemoteTarball(target, resolution, opts);
}
async function fetchFromRemoteTarball(ctx, unpackTo, dist, opts) {
    try {
        return await fetchFromLocalTarball(opts.cachedTarballLocation, unpackTo, {
            integrity: dist.integrity,
        });
    }
    catch (err) {
        // ignore errors for missing files or broken/partial archives
        switch (err.code) {
            case 'Z_BUF_ERROR':
                logger_1.storeLogger.warn(`Redownloading corrupted cached tarball: ${opts.cachedTarballLocation}`);
                break;
            case 'ENOENT':
                break;
            default:
                throw err;
        }
        if (ctx.offline) {
            throw new errorTypes_1.PnpmError('ERR_PNPM_NO_OFFLINE_TARBALL', `Could not find ${opts.cachedTarballLocation} in local registry mirror`);
        }
        const auth = dist.registry ? ctx.getCredentialsByURI(dist.registry) : undefined;
        return ctx.download(dist.tarball, opts.cachedTarballLocation, {
            auth,
            ignore: ctx.ignoreFile,
            integrity: dist.integrity,
            onProgress: opts.onProgress,
            onStart: opts.onStart,
            registry: dist.registry,
            unpackTo,
        });
    }
}
async function fetchFromLocalTarball(tarball, dir, opts) {
    const tempLocation = pathTemp(dir);
    const tarballStream = fs.createReadStream(tarball);
    try {
        const filesIndex = (await Promise.all([
            unpackStream.local(tarballStream, tempLocation, {
                ignore: opts.ignore,
            }),
            opts.integrity && ssri.checkStream(tarballStream, opts.integrity),
        ]))[0];
        return { filesIndex, tempLocation };
    }
    catch (err) {
        err.attempts = 1;
        err.resource = tarball;
        throw err;
    }
}
