// @pnpm/utils/lib/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
const getAllDependenciesFromPackage_1 = require("./getAllDependenciesFromPackage");
exports.getAllDependenciesFromPackage = getAllDependenciesFromPackage_1.default;
const getSaveType_1 = require("./getSaveType");
exports.getSaveType = getSaveType_1.default;
const normalizeRegistries_1 = require("./normalizeRegistries");
exports.normalizeRegistries = normalizeRegistries_1.default;
exports.DEFAULT_REGISTRIES = normalizeRegistries_1.DEFAULT_REGISTRIES;
const pickRegistryForPackage_1 = require("./pickRegistryForPackage");
exports.pickRegistryForPackage = pickRegistryForPackage_1.default;
const realNodeModulesDir_1 = require("./realNodeModulesDir");
exports.realNodeModulesDir = realNodeModulesDir_1.default;
const safeReadPkg_1 = require("./safeReadPkg");
exports.safeReadPackage = safeReadPkg_1.default;
exports.safeReadPackageFromDir = safeReadPkg_1.fromDir;
exports.readPackage = safeReadPkg_1.default;
__export(require("./nodeIdUtils"));
__export(require("./getWantedDependencies"));
