// @pnpm/local-resolver/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const read_package_json_1 = require("@pnpm/read-package-json");
const fs = require("graceful-fs");
const path = require("path");
const ssri = require("ssri");
const parsePref_1 = require("./parsePref");
/**
 * Resolves a package hosted on the local filesystem
 */
async function resolveLocal(wantedDependency, opts) {
    const spec = parsePref_1.default(wantedDependency.pref, opts.prefix, opts.lockfileDirectory || opts.prefix);
    if (!spec)
        return null;
    if (spec.type === 'file') {
        return {
            id: spec.id,
            normalizedPref: spec.normalizedPref,
            resolution: {
                integrity: await getFileIntegrity(spec.fetchSpec),
                tarball: spec.id,
            },
            resolvedVia: 'local-filesystem',
        };
    }
    let localPkg;
    try {
        localPkg = await read_package_json_1.default(path.join(spec.fetchSpec, 'package.json'));
    }
    catch (internalErr) {
        switch (internalErr.code) {
            case 'ENOTDIR': {
                const err = new Error(`Could not install from "${spec.fetchSpec}" as it is not a directory.`);
                err['code'] = 'ERR_PNPM_NOT_PACKAGE_DIRECTORY'; // tslint:disable-line:no-string-literal
                throw err;
            }
            case 'ENOENT': {
                const err = new Error(`Could not install from "${spec.fetchSpec}" as it does not contain a package.json file.`);
                err['code'] = 'ERR_PNPM_DIRECTORY_HAS_NO_PACKAGE_JSON'; // tslint:disable-line:no-string-literal
                throw err;
            }
            default: {
                throw internalErr;
            }
        }
    }
    return {
        id: spec.id,
        normalizedPref: spec.normalizedPref,
        package: localPkg,
        resolution: {
            directory: spec.dependencyPath,
            type: 'directory',
        },
        resolvedVia: 'local-filesystem',
    };
}
exports.default = resolveLocal;
async function getFileIntegrity(filename) {
    return (await ssri.fromStream(fs.createReadStream(filename))).toString();
}
